src/Processors/QueryPlan/FinishSortingStep.cpp

```diff
@@ -58,11 +58,14 @@ void FinishSortingStep::transformPipeline(QueryPipeline & pipeline, const BuildQ
     if (pipeline.getNumStreams() > 1)
     {
         UInt64 limit_for_merging = (need_finish_sorting ? 0 : limit);
+        bool expected_one_block = limit_for_merging && limit_for_merging < max_block_size;
         auto transform = std::make_shared<MergingSortedTransform>(
                 pipeline.getHeader(),
                 pipeline.getNumStreams(),
                 prefix_description,
-                max_block_size, limit_for_merging);
+                max_block_size,
+                limit_for_merging,
+                expected_one_block);
```

Comment:
So this is the place where this entire logic is switched on, right? `expected_one_block` is set, and `read_one_range` is also set, or how was it called... maybe we need to invent some unified name  for these flags, like `has_limit_below_one_block`.